libsignal-protocol/src/crypto/native.rs

```diff
@@ -1,17 +1,112 @@
+use aes::{Aes128, Aes192, Aes256};
+use aes_ctr::{
+    Aes128Ctr,
+    Aes192Ctr, Aes256Ctr, stream_cipher::{NewStreamCipher, SyncStreamCipher},
+};
+use block_modes::{block_padding::Pkcs7, BlockMode, Cbc};
+use hmac::{Hmac, Mac};
+use sha2::{Digest, Sha256, Sha512};
+
 use crate::{
     crypto::{Crypto, Sha256Hmac, Sha512Digest, SignalCipherType},
     errors::InternalError,
 };
-use hmac::{Hmac, Mac};
-use sha2::{Digest, Sha256, Sha512};
+
+// FWI, PKCS5 padding is a subset of PKCS7
+type Aes128Cbc = Cbc<Aes128, Pkcs7>;
+type Aes192Cbc = Cbc<Aes192, Pkcs7>;
+type Aes256Cbc = Cbc<Aes256, Pkcs7>;
 
 // Create alias for HMAC-SHA256
 type HmacSha256 = Hmac<Sha256>;
 
+#[derive(Copy, Clone, Ord, PartialOrd, PartialEq, Eq)]
+pub enum Mode {
+    Encrypt,
+    Decrypt,
+}
+
 /// Cryptography routines using native Rust crates.
 #[derive(Debug, Copy, Clone, PartialEq)]
 pub struct DefaultCrypto;
 
+impl DefaultCrypto {
+    fn crypter(
+        &self,
+        mode: Mode,
+        cipher: SignalCipherType,
+        key: &[u8],
+        iv: &[u8],
+        data: &[u8],
+    ) -> Result<Vec<u8>, InternalError> {
+        let result = match (cipher, key.len()) {
+            (SignalCipherType::AesCtrNoPadding, 16) => {
```

Comment:
@Michael-F-Bryan i need to make sure of that, 
first i used a randomly generated nonce at every call using `self.fill_random(...)` but i found that using the `IV` here is a valid `nonce`, are you agree with that ?

Reply:
To be honest, I'm not familiar enough with the nuances of AES to know whether it's okay to use the provided IV as a nonce. Can we look at the `libsignal-protocol-c` source code and check whether our `crypter()` function will be provided with bytes that came from `crypto.fill_random()`?